src/main/java/com/fauna/response/StreamEvent.java

```diff
@@ -1,48 +1,112 @@
+import static com.fauna.constants.ResponseFields.STREAM_TYPE_FIELD_NAME;
+
 public class StreamEvent<E> {
-    private final StreamEventWire wire;
-    private final Class<E> elementClass;
+    private static final Codec<QueryStats> statsCodec = DefaultCodecProvider.SINGLETON.get(QueryStats.class);
+    public enum EventType {
+        STATUS, ADD, UPDATE, REMOVE
+    }
+
+    private final EventType type;
+    private final String cursor;
+    private final Long txn_ts;
     private final E data;
+    private final QueryStats stats;
+
+
+    public StreamEvent(EventType type, String cursor, Long txn_ts, E data, QueryStats stats) {
+        this.type = type;
+        this.cursor = cursor;
+        this.txn_ts = txn_ts;
+        this.data = data;
+        this.stats = stats;
+    }
 
-    public StreamEvent(StreamEventWire wire, Class<E> elementClass) {
-        this.wire = wire;
-        this.elementClass = elementClass;
-        try {
-            Codec<E> elementCodec = DefaultCodecProvider.SINGLETON.get(elementClass);
-            if (wire.getData() != null && wire.getData() != "null") {
-                this.data = elementCodec.decode(new UTF8FaunaParser(wire.getData()));
-            } else {
-                this.data = null;
+    private static StreamEvent.EventType parseEventType(JsonParser parser) throws IOException {
+        if (parser.nextToken() == VALUE_STRING) {
+            String typeString = parser.getText().toUpperCase();
+            try {
+                return StreamEvent.EventType.valueOf(typeString);
+            } catch (IllegalArgumentException e) {
+                throw new ClientException("Invalid event type: " + typeString, e);
             }
-        } catch (IOException e) {
-            throw new ClientException("Failed to parse data.", e);
+        } else {
+            throw new ClientException("Event type should be a string, but got a " + parser.currentToken().asString());
+        }
+    }
+
+    public static <E> StreamEvent<E> parse(JsonParser parser, Codec<E> dataCodec) throws IOException {
+        if (parser.nextToken() == START_OBJECT) {
+            String cursor = null;
+            StreamEvent.EventType eventType = null;
+            QueryStats stats = null;
+            E data = null;
+            Long txn_ts = null;
+            while (parser.nextToken() == FIELD_NAME) {
+                String fieldName = parser.getValueAsString();
+                switch (fieldName) {
+                    case STREAM_CURSOR_FIELD_NAME:
+                        parser.nextToken();
+                        cursor = parser.getText();
+                        break;
+                    case DATA_FIELD_NAME:
+                        UTF8FaunaParser faunaParser = new UTF8FaunaParser(parser);
+                        if (faunaParser.getCurrentTokenType() == FaunaTokenType.NONE) {
+                            faunaParser.read();
+                        }
+                        data = dataCodec.decode(faunaParser);
```

Comment:
I didn't think that the UTF8FaunaParser knew yet how to handle with a partially consumed JsonParser that it's not supposed to exhaust. This is the reason I opted for the json object mapper with the pass-through deserializer for other cases–it felt less invasive and we could work on the ideal impl after GA.

Reply:
Discussed in 1:1... It actually works similar to the generic types like `ListOf` and `MapOf`. The only difference here is that we're going from `JsonParser -> UTF8FaunaParser` instead of `UTF8FaunaParser -> UTF8FaunaParser`. But since `UTF8FaunaParser` always has an associated `JsonParser` it's actually equivalent.